### DISCUSSION TITLE
Broadcasting processed video frame to NVR

### DISCUSSION THREAD

**Discussion by AlainPilon** [roboflow/supervision#402] [Q&A] [answered]
For my current project, I capture and process with Supervision a video feed from an IP camera and I need to broadcast the annotated output so it can be recording by an NVR in our data center. Storing the video file locally is not a viable option for us for multiple business/infra reasons.

My googling skills are failing me on this as I am not familiar with the domains terminology. Anyone has experience doing something like this? 

Thx!

**Comment by PawelPeczek-Roboflow** [comment, marked as answer]
Hello there :)
If I understand you correctly - you would like to to emit a stream of annotated frames and intercept it via NVR on the other end.
It seems like the problem is almost equivalent to the situation that you intercept camera stream in NVR. In your scenario - you need something that will emit stream of annotated frames in a format that is accepted by NVR - it can be RTSP stream or other type if you wish.
The tricky part would be to find the easiest streaming format (from implementational perspective) to be accepted by NVR. Once it is found - you need to encode the video frames and send them via stream. Sounds like a magic, but bear with me.
I've not tested that on my own, but it seems like a combination of OpenCV and ffmpeg could help. Take a look here: https://stackoverflow.com/questions/69379674/how-to-stream-cv2-videowriter-frames-to-and-rtsp-server
That may not work for various reasons (like NVR not supporting RTSP, or expecting to act as RTSP client), but let's start with that for now.

**Reply by PawelPeczek-Roboflow** [reply]
At the moment, library on its own do not support this, but maybe it will in the future

**Reply by SkalskiP** [reply]
@AlainPilon, is it enough to get you started?

**Reply by AlainPilon** [reply]
Yes @PawelPeczek-Roboflow, that is pretty much the idea and your link seems promising and I will investigate it once I am done with my current task.  

Thanks a lot!

**Comment by maddust** [comment]
@AlainPilon 

I use a combination of FFmpeg and mediamtx (formely called rstp simple server)  to broadcast the processed frame .

Mediamtx documentation here : [https://github.com/bluenviron/mediamtx](url)

you can broadcast your processed frame pretty much in any standard format including Webrtc, HLS , Rtsp, Rtmp ..... 

Below an example  of the code you need to include 

```
import subprocess as sp

def init_ffmpeg(command):
    proc = sp.Popen(command, stdin=sp.PIPE, shell=False)
    return proc     

##PLACE THIS ON YOU MAIN CODE###

stream=f"rtsp://localhost:8554/yourstream"  # <<<<<<< your broadcast stream (not your ip camera stream)
    command = [
    'ffmpeg',
    '-y',
    '-f', 'rawvideo',
    '-vcodec', 'rawvideo',
    '-pix_fmt', 'bgr24',
    '-s', '{}x{}'.format(width, height),
    '-r', str(video_info.fps),
    '-i', '-',
    '-an',
    '-c:v', 'h264_nvenc',
    '-pix_fmt', 'yuv420p',
    '-profile:v', 'baseline',  # Explicit profile for h264_nvenc
    '-preset', 'fast',
    '-maxrate', '2500k',
    '-bufsize', '500k', #'-max_muxing_queue_size', '1024', 
    '-g', '30',
    '-f', 'rtsp',
    stream
]
proc = init_ffmpeg(command)

#At the End of your frame processing  loop place this #

proc.stdin.write(frame.tobytes())<<<<< pass your frame to the subprocess

##make sure you close the subprocess ##

cv2.imshow("frame",frame)
                if cv2.waitKey(1) & 0xFF == ord("q"):
                    cap.release()
                    proc.stdin.close()
                    cv2.destroyAllWindows()
                    exit(0)  # exit the program
```

hope it helps!
